File: src/builtins/sqlite.js
// node:sqlite — synchronous SQLite, backed by the native rusqlite bridge
// (__velox_sqlite_*). Implements the Node 22 API surface: DatabaseSync +
// StatementSync (run/get/all/iterate). Values cross as JSON; BLOBs and BigInts
// are tagged ({t:'blob'|'bigint', v}) so binary data and >2^53 integers survive.

var hasNative = typeof __velox_sqlite_open === 'function';

// --- value tagging (mirror of src/sqlite.rs) -------------------------------

function encodeParam(v) {
  if (v === null || v === undefined) return null;
  var t = typeof v;
  if (t === 'number' || t === 'string') return v;
  if (t === 'boolean') return v ? 1 : 0;
  if (t === 'bigint') return { t: 'bigint', v: v.toString() };
  if (globalThis.Buffer && globalThis.Buffer.isBuffer(v)) {
    return { t: 'blob', v: v.toString('base64') };
  }
  if (v instanceof Uint8Array || (v && v.buffer instanceof ArrayBuffer && typeof v.byteLength === 'number')) {
    return { t: 'blob', v: globalThis.Buffer.from(v.buffer, v.byteOffset, v.byteLength).toString('base64') };
  }
  if (v instanceof ArrayBuffer) {
    return { t: 'blob', v: globalThis.Buffer.from(v).toString('base64') };
  }
  throw new TypeError('Unsupported SQLite parameter type: ' + t);
}

function decodeValue(v) {
  if (v !== null && typeof v === 'object' && typeof v.t === 'string') {
    if (v.t === 'blob') return globalThis.Buffer.from(v.v, 'base64');
    if (v.t === 'bigint') return BigInt(v.v);
  }
  return v;
}

// Normalize the variadic params of run/get/all/iterate into the JSON shape the
// native expects: a positional array, or a named object whose keys carry their
// sigil. Node accepts a single object of named params, otherwise positionals.
function buildParams(args, allowBareNamed) {
  if (args.length === 1 && args[0] !== null && typeof args[0] === 'object'
      && !globalThis.Buffer.isBuffer(args[0]) && !(args[0] instanceof Uint8Array)
      && !Array.isArray(args[0]) && !(args[0] instanceof ArrayBuffer)) {
    var named = {};
    var src = args[0];
    for (var k in src) {
      if (!Object.prototype.hasOwnProperty.call(src, k)) continue;
      // Allow bare names (`id`) to bind `:id`/`@id`/`$id`; default to `:`.
      var key = /^[:@$]/.test(k) ? k : (allowBareNamed ? ':' + k : k);
      named[key] = encodeParam(src[k]);
    }
    return JSON.stringify(named);
  }
  var positional = [];
  for (var i = 0; i < args.length; i++) positional.push(encodeParam(args[i]));
  return JSON.stringify(positional);
}

function rowsToObjects(parsed) {
  var cols = parsed.columns;
  var out = [];
  for (var r = 0; r < parsed.rows.length; r++) {
    var row = parsed.rows[r];
    var obj = {};
    for (var c = 0; c < cols.length; c++) obj[cols[c]] = decodeValue(row[c]);
    out.push(obj);
  }
  return out;
}

// --- StatementSync ---------------------------------------------------------

function StatementSync(dbId, sql, db) {
  this._dbId = dbId;
  this._sql = sql;
  this._db = db;
  this._bigints = false;
  // Node allows bare named-parameter keys (`id` binds `:id`) by default.
  this._bareNamed = true;
  this.sourceSQL = sql;
}

StatementSync.prototype.run = function () {
  var params = buildParams(arguments, this._bareNamed);
  var res = JSON.parse(__velox_sqlite_run(this._dbId, this._sql, params));
  return { changes: res.changes, lastInsertRowid: decodeValue(res.lastInsertRowid) };
};

StatementSync.prototype.all = function () {
  var params = buildParams(arguments, this._bareNamed);
  var parsed = JSON.parse(__velox_sqlite_query(this._dbId, this._sql, params, this._bigints));
  return rowsToObjects(parsed);
};

StatementSync.prototype.get = function () {
  var params = buildParams(arguments, this._bareNamed);
  var parsed = JSON.parse(__velox_sqlite_query(this._dbId, this._sql, params, this._bigints));
  if (parsed.rows.length === 0) return undefined;
  return rowsToObjects({ columns: parsed.columns, rows: [parsed.rows[0]] })[0];
};

StatementSync.prototype.iterate = function () {
  var rows = this.all.apply(this, arguments);
  var i = 0;
  var it = {
    next: function () {
      return i < rows.length ? { value: rows[i++], done: false } : { value: undefined, done: true };
    },
    return: function () { i = rows.length; return { value: undefined, done: true }; },
  };
  it[Symbol.iterator] = function () { return this; };
  return it;
};

StatementSync.prototype.columns = function () {
  // Run with no rows to learn the column names without scanning data.
  var parsed = JSON.parse(__velox_sqlite_query(this._dbId, this._sql + ' LIMIT 0', '[]', false));
  return parsed.columns.map(function (name) {
    return { column: name, name: name, table: null, database: null, type: null };
  });
};

StatementSync.prototype.setReadBigInts = function (flag) { this._bigints = !!flag; return this; };
StatementSync.prototype.setAllowBareNamedParameters = function (flag) { this._bareNamed = !!flag; return this; };
StatementSync.prototype.setAllowUnknownNamedParameters = function () { return this; };

// --- DatabaseSync ----------------------------------------------------------

function DatabaseSync(path, options) {
  if (!hasNative) throw new Error('node:sqlite is not available in this build');
  this._path = path === undefined || path === null ? ':memory:' : String(path);
  this._id = null;
  this._open = false;
  options = options || {};
  // `open: false` defers opening until .open() is called.
  if (options.open !== false) this.open();
}

DatabaseSync.prototype.open = function () {
  if (this._open) throw new Error('database is already open');
  this._id = __velox_sqlite_open(this._path);
  this._open = true;
};

DatabaseSync.prototype.close = function () {
  if (!this._open) return;
  __velox_sqlite_close(this._id);
  this._open = false;
  this._id = null;
};

DatabaseSync.prototype.exec = function (sql) {
  if (!this._open) throw new Error('database is not open');
  __velox_sqlite_exec(this._id, String(sql));
};

DatabaseSync.prototype.prepare = function (sql) {
  if (!this._open) throw new Error('database is not open');
  return new StatementSync(this._id, String(sql), this);
};

DatabaseSync.prototype.location = function () {
  return this._path === ':memory:' ? null : this._path;
};

// Custom functions/aggregates aren't bridged yet; surface a clear error rather
// than silently no-op so callers know.
DatabaseSync.prototype.function = function () {
  throw new Error('DatabaseSync.function() is not supported in velox yet');
};
DatabaseSync.prototype.aggregate = function () {
  throw new Error('DatabaseSync.aggregate() is not supported in velox yet');
};

Object.defineProperty(DatabaseSync.prototype, 'isOpen', {
  get: function () { return this._open; },
  configurable: true,
});

DatabaseSync.prototype[Symbol.dispose] = function () { this.close(); };

var constants = {
  SQLITE_CHANGESET_OMIT: 0,
  SQLITE_CHANGESET_REPLACE: 1,
  SQLITE_CHANGESET_ABORT: 2,
};

module.exports = {
  DatabaseSync: DatabaseSync,
  StatementSync: StatementSync,
  constants: constants,
};
module.exports.default = module.exports;

File: src/sqlite.rs
//! `node:sqlite` — synchronous embedded SQLite via `rusqlite` (SQLite compiled
//! in via the `bundled` feature, so there's no system dependency). The JS shim
//! (`src/builtins/sqlite.js`) wraps these natives in `DatabaseSync` /
//! `StatementSync`.
//!
//! Values cross the FFI as JSON. BLOBs and integers that don't fit a JS number
//! are encoded as tagged objects — `{"t":"blob","v":"<base64>"}` and
//! `{"t":"bigint","v":"<decimal>"}` — so binary data and >2^53 integers survive
//! the JSON hop; the shim reconstitutes them as `Buffer`/`BigInt`. Statements are
//! re-prepared from their SQL on each call (rusqlite's prepare is cheap and this
//! sidesteps holding a borrow of the connection across the JS boundary).
//!
//! Connections live in a thread-local registry keyed by id, so each worker
//! thread (which has its own JS context) gets an independent set.

use std::cell::RefCell;
use std::collections::HashMap;
use std::ptr;

use base64::Engine;
use objc2_javascript_core::{JSContextRef, JSObjectRef, JSValue, JSValueRef};
use rusqlite::Connection;
use rusqlite::types::{Value, ValueRef};
use serde_json::{Value as Json, json};

use crate::event_loop::{arg_slice, register};
use crate::node::{call_named, js_string};
use crate::runtime::js_value_to_string;

thread_local! {
    static DBS: RefCell<HashMap<u64, Connection>> = RefCell::new(HashMap::new());
    static NEXT_ID: RefCell<u64> = const { RefCell::new(1) };
}

const B64: base64::engine::GeneralPurpose = base64::engine::general_purpose::STANDARD;

/// The largest integer magnitude that round-trips exactly through an f64.
const MAX_SAFE_INT: i64 = 9_007_199_254_740_991;

/// Register the native SQLite functions.
pub fn install(ctx: JSContextRef) {
    unsafe {
        register(ctx, c"__velox_sqlite_open", open_fn);
        register(ctx, c"__velox_sqlite_close", close_fn);
        register(ctx, c"__velox_sqlite_exec", exec_fn);
        register(ctx, c"__velox_sqlite_run", run_fn);
        register(ctx, c"__velox_sqlite_query", query_fn);
    }
}

unsafe fn throw(ctx: JSContextRef, exception: *mut JSValueRef, message: &str) -> JSValueRef {
    unsafe {
        let args = [js_string(ctx, "ERR_SQLITE_ERROR"), js_string(ctx, message)];
        let error = call_named(ctx, c"__velox_fs_error", &args);
        if !exception.is_null() {
            *exception = error;
        }
        JSValue::new_undefined(ctx)
    }
}

fn arg_string(ctx: JSContextRef, args: &[JSValueRef], i: usize) -> String {
    args.get(i)
        .map(|v| unsafe { js_value_to_string(ctx, *v) })
        .unwrap_or_default()
}

fn arg_u64(ctx: JSContextRef, args: &[JSValueRef], i: usize) -> u64 {
    args.get(i)
        .map(|v| unsafe { JSValue::to_number(ctx, *v, ptr::null_mut()) })
        .filter(|n| n.is_finite() && *n >= 0.0)
        .map(|n| n as u64)
        .unwrap_or(0)
}

/// Convert a JSON parameter value into a SQLite value. Tagged objects carry
/// blobs/bigints; plain JSON numbers become INTEGER when integral, else REAL.
fn json_to_sql(v: &Json) -> Value {
    match v {
        Json::Null => Value::Null,
        Json::Bool(b) => Value::Integer(*b as i64),
        Json::Number(n) => {
            if let Some(i) = n.as_i64() {
                Value::Integer(i)
            } else if let Some(f) = n.as_f64() {
                Value::Real(f)
            } else {
                Value::Null
            }
        }
        Json::String(s) => Value::Text(s.clone()),
        Json::Object(o) => match o.get("t").and_then(|t| t.as_str()) {
            Some("blob") => {
                let v = o.get("v").and_then(|x| x.as_str()).unwrap_or("");
                Value::Blob(B64.decode(v).unwrap_or_default())
            }
            Some("bigint") => {
                let v = o.get("v").and_then(|x| x.as_str()).unwrap_or("0");
                Value::Integer(v.parse().unwrap_or(0))
            }
            _ => Value::Null,
        },
        // Arrays aren't valid bind values.
        Json::Array(_) => Value::Null,
    }
}

/// Convert a SQLite column value into JSON for the JS side. Integers outside the
/// JS safe range (or when `bigints` is set) come back bigint-tagged.
fn sql_to_json(v: ValueRef, bigints: bool) -> Json {
    match v {
        ValueRef::Null => Json::Null,
        ValueRef::Integer(i) => {
            if bigints || i > MAX_SAFE_INT || i < -MAX_SAFE_INT {
                json!({ "t": "bigint", "v": i.to_string() })
            } else {
                json!(i)
            }
        }
        ValueRef::Real(f) => json!(f),
        ValueRef::Text(t) => json!(String::from_utf8_lossy(t).into_owned()),
        ValueRef::Blob(b) => json!({ "t": "blob", "v": B64.encode(b) }),
    }
}

/// Bind a parsed parameter set (JSON array → positional, JSON object → named) to
/// a prepared statement.
fn bind_params(stmt: &mut rusqlite::Statement, params: &Json) -> Result<(), rusqlite::Error> {
    match params {
        Json::Array(arr) => {
            for (i, v) in arr.iter().enumerate() {
                stmt.raw_bind_parameter(i + 1, json_to_sql(v))?;
            }
        }
        Json::Object(obj) => {
            for (k, v) in obj {
                // The shim sends names with their sigil (`:name`/`@name`/`$name`).
                if let Some(idx) = stmt.parameter_index(k)? {
                    stmt.raw_bind_parameter(idx, json_to_sql(v))?;
                }
            }
        }
        _ => {}
    }
    Ok(())
}

/// `__velox_sqlite_open(path)` → numeric connection id. `:memory:`/empty opens an
/// in-memory database.
unsafe extern "C-unwind" fn open_fn(
    ctx: JSContextRef,
    _f: JSObjectRef,
    _t: JSObjectRef,
    argc: usize,
    argv: *mut JSValueRef,
    exc: *mut JSValueRef,
) -> JSValueRef {
    let args = arg_slice(argc, argv);
    let path = arg_string(ctx, args, 0);
    let conn = if path.is_empty() || path == ":memory:" {
        Connection::open_in_memory()
    } else {
        Connection::open(&path)
    };
    match conn {
        Ok(c) => {
            let id = NEXT_ID.with(|n| {
                let mut n = n.borrow_mut();
                let id = *n;
                *n += 1;
                id
            });
            DBS.with(|d| d.borrow_mut().insert(id, c));
            unsafe { JSValue::new_number(ctx, id as f64) }
        }
        Err(e) => unsafe { throw(ctx, exc, &e.to_string()) },
    }
}

/// `__velox_sqlite_close(id)` — close and drop the connection.
unsafe extern "C-unwind" fn close_fn(
    ctx: JSContextRef,
    _f: JSObjectRef,
    _t: JSObjectRef,
    argc: usize,
    argv: *mut JSValueRef,
    _exc: *mut JSValueRef,
) -> JSValueRef {
    let args = arg_slice(argc, argv);
    let id = arg_u64(ctx, args, 0);
    DBS.with(|d| d.borrow_mut().remove(&id));
    unsafe { JSValue::new_undefined(ctx) }
}

/// `__velox_sqlite_exec(id, sql)` — run a batch of statements, no results.
unsafe extern "C-unwind" fn exec_fn(
    ctx: JSContextRef,
    _f: JSObjectRef,
    _t: JSObjectRef,
    argc: usize,
    argv: *mut JSValueRef,
    exc: *mut JSValueRef,
) -> JSValueRef {
    let args = arg_slice(argc, argv);
    let id = arg_u64(ctx, args, 0);
    let sql = arg_string(ctx, args, 1);
    let result = DBS.with(|d| {
        d.borrow()
            .get(&id)
            .ok_or_else(|| "database is not open".to_string())
            .and_then(|c| c.execute_batch(&sql).map_err(|e| e.to_string()))
    });
    match result {
        Ok(()) => unsafe { JSValue::new_undefined(ctx) },
        Err(e) => unsafe { throw(ctx, exc, &e) },
    }
}

/// `__velox_sqlite_run(id, sql, paramsJson)` → JSON `{changes, lastInsertRowid}`.
unsafe extern "C-unwind" fn run_fn(
    ctx: JSContextRef,
    _f: JSObjectRef,
    _t: JSObjectRef,
    argc: usize,
    argv: *mut JSValueRef,
    exc: *mut JSValueRef,
) -> JSValueRef {
    let args = arg_slice(argc, argv);
    let id = arg_u64(ctx, args, 0);
    let sql = arg_string(ctx, args, 1);
    let params: Json = serde_json::from_str(&arg_string(ctx, args, 2)).unwrap_or(Json::Null);

    let result = DBS.with(|d| -> Result<String, String> {
        let dbs = d.borrow();
        let conn = dbs.get(&id).ok_or("database is not open")?;
        let mut stmt = conn.prepare(&sql).map_err(|e| e.to_string())?;
        bind_params(&mut stmt, &params).map_err(|e| e.to_string())?;
        let changes = stmt.raw_execute().map_err(|e| e.to_string())?;
        let last = conn.last_insert_rowid();
        let last_json = if last > MAX_SAFE_INT || last < -MAX_SAFE_INT {
            json!({ "t": "bigint", "v": last.to_string() })
        } else {
            json!(last)
        };
        Ok(json!({ "changes": changes, "lastInsertRowid": last_json }).to_string())
    });
    match result {
        Ok(s) => unsafe { js_string(ctx, &s) },
        Err(e) => unsafe { throw(ctx, exc, &e) },
    }
}

/// `__velox_sqlite_query(id, sql, paramsJson, bigints)` → JSON
/// `{columns: [...], rows: [[...], ...]}`. The shim turns this into objects and
/// picks the first row for `.get()` / all rows for `.all()` / an iterator for
/// `.iterate()`.
unsafe extern "C-unwind" fn query_fn(
    ctx: JSContextRef,
    _f: JSObjectRef,
    _t: JSObjectRef,
    argc: usize,
    argv: *mut JSValueRef,
    exc: *mut JSValueRef,
) -> JSValueRef {
    let args = arg_slice(argc, argv);
    let id = arg_u64(ctx, args, 0);
    let sql = arg_string(ctx, args, 1);
    let params: Json = serde_json::from_str(&arg_string(ctx, args, 2)).unwrap_or(Json::Null);
    let bigints = args
        .get(3)
        .map(|v| unsafe { JSValue::to_boolean(ctx, *v) })
        .unwrap_or(false);

    let result = DBS.with(|d| -> Result<String, String> {
        let dbs = d.borrow();
        let conn = dbs.get(&id).ok_or("database is not open")?;
        let mut stmt = conn.prepare(&sql).map_err(|e| e.to_string())?;
        let columns: Vec<String> = stmt.column_names().iter().map(|s| s.to_string()).collect();
        let col_count = columns.len();
        bind_params(&mut stmt, &params).map_err(|e| e.to_string())?;
        let mut rows = stmt.raw_query();
        let mut out: Vec<Json> = Vec::new();
        while let Some(row) = rows.next().map_err(|e| e.to_string())? {
            let mut cells: Vec<Json> = Vec::with_capacity(col_count);
            for i in 0..col_count {
                let v = row.get_ref(i).map_err(|e| e.to_string())?;
                cells.push(sql_to_json(v, bigints));
            }
            out.push(Json::Array(cells));
        }
        Ok(json!({ "columns": columns, "rows": out }).to_string())
    });
    match result {
        Ok(s) => unsafe { js_string(ctx, &s) },
        Err(e) => unsafe { throw(ctx, exc, &e) },
    }
}
